### PR TITLE
feat(AAPD-148): base class for all validators

### PR DESCRIPTION
**Pull request by suityou01**
## Ticket Number

https://pins-ds.atlassian.net/browse/AAPD-148

## Description of change

Abstract base class for all validators

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `main` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/forms-web-app/src/dynamic-forms/validator/baseValidator.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @abstract
+ * @class BaseValidator
+ */
+class BaseValidator {
+	constructor() {
+		if (this.constructor == BaseValidator) {
+			throw new Error("Abstract classes can't be instantiated.");
```
Comment (chrishawldev) [inline packages/forms-web-app/src/dynamic-forms/validator/baseValidator.js:8]: Can we get a unit test on this please, looks like a nice enhancement good shout

**Review by chrishawldev — COMMENTED** [review]
Do you think it's worth implementing the type utility method on this class
class BaseValidator {
	// The type of the vlaidator
	type;

}
Or do you think it's just as easy to do it inline / via utility method in the validate middleware here https://github.com/Planning-Inspectorate/appeal-planning-decision/blob/spike/dynamic-questionnaire-renderer-validation-examples/packages/questionnaire-web-app/src/validator/validator.js#L14 when we tie it all together?

**Review by chrishawldev — COMMENTED** [review]
Only other thing is this discussion point https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/3009#pullrequestreview-1580600723, we don't necessarily need to change anything but need to capture our decision point and we can add it to one of the other validation tickets if we're looking to stick it on the middleware, either way works for me so I'll leave it your discretion

**Comment by suityou01** [comment]
> Do you think it's worth implementing the type utility method on this class class BaseValidator { // The type of the vlaidator type;
> 
> } Or do you think it's just as easy to do it inline / via utility method in the validate middleware here https://github.com/Planning-Inspectorate/appeal-planning-decision/blob/spike/dynamic-questionnaire-renderer-validation-examples/packages/questionnaire-web-app/src/validator/validator.js#L14 when we tie it all together?

I don't think it's needed? Is the "instanceof" keyword not sufficient?

**Comment by chrishawldev** [comment]
> > Do you think it's worth implementing the type utility method on this class class BaseValidator { // The type of the vlaidator type;
> > } Or do you think it's just as easy to do it inline / via utility method in the validate middleware here https://github.com/Planning-Inspectorate/appeal-planning-decision/blob/spike/dynamic-questionnaire-renderer-validation-examples/packages/questionnaire-web-app/src/validator/validator.js#L14 when we tie it all together?
> 
> I don't think it's needed? Is the "instanceof" keyword not sufficient?

Cool that works for me. I'll add a comment to the Jira ticket. Nice one

**Review by chrishawldev — APPROVED** [review]

**Pull request merged**